mtp_api/apps/core/excel.py

```diff
@@ -0,0 +1,141 @@
+    def open_workbook(cls, file: File):
+        extension = (file.name or '').rsplit('.')[-1].lower()
+        try:
+            return cls.subclasses[extension](file)
+        except KeyError:
+            raise TypeError('Unknown file extension')
+
+    def __init__(self, file: File):
+        self.file = file
+
+    def __enter__(self):
+        # subclasses open file here
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # subclasses close file here
+        raise NotImplementedError
+
+    def get_sheet(self, index: int) -> 'ExcelWorksheet':
+        return ExcelWorksheet.subclasses[self.extension](self, index)
+
+
+class XLSWorkbook(ExcelWorkbook):
+    extension = 'xls'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        from xlrd.book import Book
+
+        self._workbook: Book
+
+    def __enter__(self):
+        from xlrd import open_workbook, XLRDError
+        from xlrd.compdoc import CompDocError
+
+        try:
+            self._workbook = open_workbook(
+                filename=self.file.name,
+                file_contents=self.file.read(),
+                on_demand=True,
+            ).__enter__()
+        except (XLRDError, CompDocError) as e:
+            raise TypeError('Cannot load XLS file') from e
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return self._workbook.__exit__(exc_type, exc_val, exc_tb)
+
+    def get_sheet(self, index: int) -> 'XLSWorksheet':
+        return XLSWorksheet(self, index)
+
+
+class XLSXWorkbook(ExcelWorkbook):
+    extension = 'xlsx'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        from openpyxl.workbook import Workbook
+
+        self._workbook: Workbook
+
+    def __enter__(self):
+        from openpyxl import load_workbook
+
+        try:
+            self._workbook = load_workbook(self.file, read_only=True)
+        except ValueError as e:
+            raise TypeError('Cannot load XLSX file') from e
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._workbook.close()
+
+    def get_sheet(self, index: int) -> 'XLSXWorksheet':
+        return XLSXWorksheet(self, index)
+
+
+class ExcelWorksheet:
+    """
+    Returned by ExcelWorkbook.get_sheet(index),
+    a common interface for dealing with XLS and XLSX worksheets
+    """
+    subclasses = {}
+    extension = NotImplemented
+
+    def __init_subclass__(cls):
+        cls.subclasses[cls.extension] = cls
+
+    @property
+    def row_count(self) -> int:
+        raise NotImplementedError
+
+    def cell_value(self, row: int, column: int):
+        raise NotImplementedError
+
+
+class XLSWorksheet(ExcelWorksheet):
+    extension = 'xls'
+
+    def __init__(self, workbook: XLSWorkbook, index: int):
+        self.workbook = workbook
+        self.index = index
+        self._worksheet = self.workbook._workbook.get_sheet(index)
```

Comment:
As discussed offline, I think it's worth playing with the idea of getting the concrete sheet in the `WorkBook`s classes as that would avoid the need of accessing another class "private" properties. And I think it's fair for the Workbook class to do that and pass it to the sheet class.

(Oh I see you're already playing with it ❤️ 👍)